### DISCUSSION TITLE
Custom configurations management

### DISCUSSION THREAD

**Discussion by sestegra** [vmware/packer-examples-for-vsphere#50] [Ideas]
Hello,

To ease the contribution, avoid to push sensitive values in a PR and support multiple configurations, I suggest to modify a bit how Packer variables are managed.

### Suggestion 1 - Copy `builds/*pkrvars.hcl` to `custom` folder
Instead of editing variables in-place (Git version control), I suggest to create Git ignored folder `custom`
So, a global instruction to do the copy would be added
```
mkdir -p custom
cp builds/*pkrvars.hcl custom
```

Then, following instructions 

> Edit the `/builds/*.pkvars.hcl` file to configure the following:

would become

> Edit the `/custom/*.pkvars.hcl` file to configure the following:

Finally, add a variable to `build.sh` to point into a customized folder.
This variable should be used for different customized folders.

```
./build.sh custom
./build.sh custom-alice
./build.sh custom-bob
```

### Suggestion 2 - Git ignore `certificates` folder
To avoid to push sensitive values in a PR, I suggest to remove `certificates` from Git version control.

### Suggestion 3 - Create and Git ignore `ssh` folder
Instead of setting SSH public keys (`builder` and `ansible`) directly in the related variable, I suggest to use `file` function to load from a file located in a dedicated folder (e.g. `ssh`).

### Suggestion 4 - Use skaffold folder for customization
```
my-custom-variables
├── builds
│   ├── ansible.pkrvars.hcl
│   ├── build.pkrvars.hcl
│   ├── common.pkrvars.hcl
│   ├── proxy.pkrvars.hcl
│   ├── rhsm.pkrvars.hcl
│   └── vsphere.pkrvars.hcl
├── certificates
└── ssh
```

Then, use the `path` variable to manage customized values in build sources.
See documentation here: https://www.packer.io/docs/templates/hcl_templates/path-variables

What do you think about it?

Kind regards,
Stéphane

**Comment by tenthirtyam** [comment]
This would be a very nice modification. Please clone to an enhancement and we can continue to discuss further here.

Ryan

**Comment by sestegra** [comment]
Suggestion 1  is implemented in #57.

**Reply by tenthirtyam** [reply]
Merged https://github.com/rainpole/packer-vsphere/pull/57.

Ryan

**Comment by tenthirtyam** [comment]
For Suggestion 3 (re: the public keys) I do prefer the option to pass these as sensitive input variables. In a prior release these were files in a `keys` directory.

Ryan

**Reply by sestegra** [reply]
For Suggestion 3, no change is required, except README.md to explain the alternative that I suggested.
Now, as `config` folder is private, the configuration is more flexible.

So, both following configurations are possible, and the variable is kept sensitive in both case.
- in-place public key
```
build_key = "ssh-rsa xxx"
```
- from file
```
build_key = file("${path.root}/ssh/build_id_rsa.pub")
```

**Reply by tenthirtyam** [reply]
Ah, good point. Much cleaner.

**Reply by tenthirtyam** [reply]
I'll take a look at updating the README this week.

Ryan

**Comment by tenthirtyam** [comment]
For suggestion 2, certificates, it may be best to have a `.gitkeep` so the folder is included in a clone, but have a `.gitignore` for `.crt` and `.pb7` files. 

In addition (already planned) change the included certificates as `.crt.example` and `.pb7.example`.

Ryan

**Reply by sestegra** [reply]
For Suggestion 2, after reading https://github.com/rainpole/packer-vsphere/issues/54, I prefer to do the certificate(s) deployment with such solution.

On top of this, the certificates could be different per `config`.

**Comment by sestegra** [comment]
I will work on a PR to have `path.cwd` different from `path.root` when using `build.sh`.
https://www.packer.io/docs/templates/hcl_templates/path-variables

Then, I will do following updates:
- replace `path.cwd` by `path.root` to access files `data` directory
- replace `../../certificates` by `${path.cwd}/certificates` to access `certificates`
- replace `../../manifests` by `${path.cwd}/manifests` to access `manifests`
- replace `../../scripts` by `${path.cwd}/scripts` to access `scripts`

My goal is to support launching `packer` from whatever directory, store `certificates` and `manifests` in that directory (from a different script than `build.sh`).
